### PR TITLE
Add C# 9 test cases - rules A* ... B*

### PR DESCRIPTION
**Pull request by andrei-epure-sonarsource**
Related to #3668

### CODE REVIEW

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/MetadataReferences/NuGetMetadataReference.cs [hunk 1]
```diff
@@ -28,69 +28,69 @@ internal static class NuGetMetadataReference
     {
         // hardcoded version
 
-        public static References MicrosoftVisualStudioQualityToolsUnitTestFramework => Create("VS.QualityTools.UnitTestFramework", "15.0.27323.2");
-        public static References MSTestTestFrameworkV1 => Create("MSTest.TestFramework", "1.1.11");
+        public static References MicrosoftVisualStudioQualityToolsUnitTestFramework => CreateWithCommandLine("VS.QualityTools.UnitTestFramework", "15.0.27323.2");
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/MetadataReferences/NuGetMetadataReference.cs:31]: I had local failures... this is why I did this commit

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/AbstractTypesShouldNotHaveConstructorsTests.cs [hunk 2]
```diff
@@ -34,5 +34,14 @@ public void NoPublicConstructor()
         {
             Verifier.VerifyAnalyzer(@"TestCases\AbstractTypesShouldNotHaveConstructors.cs", new AbstractTypesShouldNotHaveConstructors());
         }
+
+        [TestMethod]
+        [TestCategory("Rule")]
+        public void NoPublicConstructor_TopLevelStatements()
+        {
+            Verifier.VerifyAnalyzer(@"TestCases\AbstractTypesShouldNotHaveConstructors.TopLevelStatements.cs", new AbstractTypesShouldNotHaveConstructors(),
+                options: ParseOptionsHelper.FromCSharp9, checkMode: CompilationErrorBehavior.Ignore);
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/AbstractTypesShouldNotHaveConstructorsTests.cs:43]: there's a compilation error - the compiler says it must be inside an executable project

**Comment by andrei-epure-sonarsource** [comment]
I need to separate pre-C# 9 and C# 9 examples

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/AsyncVoidMethod.cs [hunk 3]
```diff
@@ -1,18 +1,31 @@
...
+            await Task.Run(() => Console.WriteLine("test"));
+        }
+
+        async void MyMethod(object sender, EventArgs args)
+        {
+            await Task.Run(() => Console.WriteLine("test"));
+        }
 
-        async void MyMethod1(object o, EventArgs e) { }
+        async void MyMethod1(object o, EventArgs e)
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/AsyncVoidMethod.cs:20]: Isn't `MyMethod1` almost the same with `MyMethod` and just parameter names changed? Can we remove it?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/AsyncVoidMethod.cs [hunk 4]
```diff
@@ -32,7 +46,13 @@ public class UwpCases
         // See issue https://github.com/SonarSource/sonar-dotnet/issues/704
         private interface ISuspendingEventArgs { }
 
-        async void MyOtherMethod1(object o, ISuspendingEventArgs args) { }
-        private async void OnSuspending(object sender, ISuspendingEventArgs e) { }
+        async void MyOtherMethod1(object o, ISuspendingEventArgs args)
+        {
+            await Task.Run(() => Console.WriteLine("test"));
+        }
+        private async void OnSuspending(object sender, ISuspendingEventArgs e)
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/AsyncVoidMethod.cs:53]: pedantic: new line is missing

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/AesManagedShouldBeWithSecureModeTest.cs [hunk 5]
```diff
@@ -35,5 +35,14 @@ public class AesManagedShouldBeWithSecureModeTest
             Verifier.VerifyAnalyzer(@"TestCases\AesManagedShouldBeWithSecureMode.cs",
                                     new AesManagedShouldBeWithSecureMode(),
                                     additionalReferences: MetadataReferenceFacade.GetSystemSecurityCryptography());
+
+        [TestMethod]
+        [TestCategory("Rule")]
+        public void AesManagedShouldBeWithSecureMode_TargetTyped() =>
+            Verifier.VerifyAnalyzer(@"TestCases\AesManagedShouldBeWithSecureMode.TargetTyped.cs",
```
[the remark this answers is not in the document]
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/AesManagedShouldBeWithSecureModeTest.cs:42]: in general, it should be CSharp9, if it's only one added file
if there are multiple, then I differentiate by mentioning what's tested more specifically

I'll make the changes, for this one I should have used CSharp9

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/AbstractTypesShouldNotHaveConstructors.Records.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/AbstractTypesShouldNotHaveConstructors.Records.cs:1]: All using statements can be removed.

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM! Just 2 more minor comments left.

**Pull request merged**